lib/moonshot/artifact_repository/s3_bucket_via_github_releases.rb

```diff
@@ -70,13 +77,138 @@ def transfer_release_asset_to_s3(version)
     def github_to_s3(version, s3_name)
       Dir.mktmpdir('github_to_s3', Dir.getwd) do |tmpdir|
         Dir.chdir(tmpdir) do
-          sh_out("hub release download #{version}")
-          file = Dir.glob("*#{version}*.tar.gz").fetch(0)
+          file = download_from_github(version)
           upload_to_s3(file, s3_name)
         end
       end
     end
 
+    # Uploads the file to s3 and verifies the checksum.
+    #
+    # @param file [String] File to be uploaded to s3.
+    # @param key [String] Name of the object to be created on s3.
+    # @raise [RuntimeError] If the file fails to upload correctly after 3
+    #                       attempts.
+    def upload_to_s3(file, key)
+      attempts = 0
+      begin
+        super
+
+        unless (checksum = checksum_file(file)).nil?
+          verify_s3_checksum(key, checksum, attempt: attempts)
+        end
+      rescue => e
```

Comment:
I'd prefer to see us rescue specific exceptions, but I'm OK with this if that isn't straight-forward (eg. because of the wide variety of possible error conditions).